python/ray/train/callbacks/logging.py

```diff
@@ -128,56 +117,13 @@ def start_training(self, logdir: str, **info):
             json.dump([], f, cls=SafeFallbackEncoder)
 
     def handle_result(self, results: List[Dict], **info):
-        if self._workers_to_log is None or results is None:
-            results_to_log = results
-        else:
-            results_to_log = [
-                result for i, result in enumerate(results)
-                if i in self._workers_to_log
-            ]
         with open(self._log_path, "r+") as f:
             loaded_results = json.load(f)
             f.seek(0)
-            json.dump(
-                loaded_results + [results_to_log], f, cls=SafeFallbackEncoder)
-
-
-class TrainingSingleWorkerLoggingCallback(TrainingLogdirMixin,
-                                          TrainingCallback, abc.ABC):
-    """Abstract Train logging callback class.
-
-    Allows only for single-worker logging.
+            json.dump(loaded_results + [results], f, cls=SafeFallbackEncoder)
```

Comment:
Unrelated to this PR, but I'm wondering if this scales well, especially if we're logging many results or a long training run. Deserializing and re-serializing all results seems like a waste. As an alternative we could seek from EOF to the last `]` and write the serialized new results there.

Reply:
Yeah agreed, I don't think this will scale. This was discussed originally [here](https://github.com/ray-project/ray/pull/17619/files#r685342190), but I think it's worthwhile to revisit this now.

1. [Tune](https://github.com/ray-project/ray/blob/master/python/ray/tune/logger.py#L500-L505) just appends individual results at the end of the file. Even though this technically isn't valid JSON, tools like `jq` can still read the file.
2. Train currently always produces a valid JSON list, although as you've mentioned the execution time will grow as the file size grows.

I took another quick look and couldn't find any existing utilities that do this (the general recommendation is to load/append/dump), but if we're doing this internally we should be able to just create our own util. I can open a GitHub issue to track this!

Reply:
Issue created here: https://github.com/ray-project/ray/issues/21416